### ISSUE TITLE
Crossentropy inconsistent results depending on tensor order

### ISSUE THREAD

**Issue by Jobanan** [pytorch/pytorch#39134]
## 🐛 Bug

While implementing contrastive loss, I discovered that depending on the order of the elements in the input tensor, torch.nn.functional.cross_entropy may return different results. With large batch size, as used for contrastive loss, this difference may reach more than 1e-2.

## To Reproduce

Short example to see the behavior where both tensor contains the same elements but in different order (except element 0 which is the "True" element):
```
print(torch.nn.functional.cross_entropy(torch.Tensor([[-0.02199082262814044952392578125, 0.050022132694721221923828125, 0.00222884747199714183807373046875, -0.122285909950733184814453125, -0.055452823638916015625]]), torch.LongTensor([0])))

print(torch.nn.functional.cross_entropy(torch.Tensor([[-0.02199082262814044952392578125, -0.122285909950733184814453125, 0.050022132694721221923828125, -0.055452823638916015625, 0.00222884747199714183807373046875]]), torch.LongTensor([0])))
```
Result:
```
tensor(1.603590130805969238281250)
tensor(1.603590250015258789062500)
```
Here the difference is small (1e-7) but with larger batch size, it may reach more than 1e-2 .

## Expected behavior

The value should not depend on the order of the element in the tensor (except for the "True" element) or at least should not reach so large values.

## Environment

 - PyTorch Version (e.g., 1.0): 1.5.0
 - OS (e.g., Linux): Ubuntu 18.04.4 LTS
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8.2
 - CUDA/cuDNN version: Run on CPU
 - GPU models and configuration: Run on CPU
 - Any other relevant information:

**Comment by vadimkantorov** [comment]
Maybe it's about numerical effects related to order of adding elements in the tensor (if elements have vastly different magnitudes)? You may try to check this by using `reduction='none'` and doing sum manually

**Comment by Jobanan** [comment]
> Maybe it's about numerical effects related to order of adding elements in the tensor (if elements have vastly different magnitudes)? You may try to check this by using `reduction='none'` and doing sum manually

Even when the batch size is only 1 (like in the example), there is a difference for the result, so I do not think that reduction could solve the problem.

**Comment by SsnL** [comment]
the order of summing will definitely affect results

**Comment by albanD** [comment]
As mentioned above, unfortunately, operations on floats are not associative. So changing their order in the input will change the result. Here 1e-7 is right at the floating point precision limit, so exactly the error we expect to see.

When you have a very large batch, this error will increase. This is expected I'm afraid. You can get the exact same behavior when you just do `sum()` with any library.

Can you share your example where you have large different > 1e-3?
